redis/src/cluster_async/connections_container.rs

```diff
@@ -0,0 +1,564 @@
+                .collect(),
+            slot_map,
+        }
+    }
+
+    fn lookup(&self, route: &Route) -> Option<ConnectionAndIdentifier<Connection>> {
+        let address = self.slot_map.slot_addr_for_route(route)?;
+        self.connection_for_address(address)
+    }
+
+    pub(crate) fn connection_for_route(
+        &self,
+        route: &Route,
+    ) -> Option<ConnectionAndIdentifier<Connection>> {
+        self.lookup(route).or_else(|| {
+            if route.slot_addr() == SlotAddr::Replica {
+                self.lookup(&Route::new(route.slot(), SlotAddr::Master))
+            } else {
+                None
+            }
+        })
+    }
+
+    pub(crate) fn all_node_connections(
+        &self,
+    ) -> impl Iterator<Item = ConnectionAndIdentifier<Connection>> + '_ {
+        self.connection_map
+            .iter()
+            .map(|(identifier, connection)| (identifier.clone(), connection.clone()))
+    }
+
+    pub(crate) fn all_primary_connections(
+        &self,
+    ) -> impl Iterator<Item = ConnectionAndIdentifier<Connection>> + '_ {
+        self.slot_map
+            .addresses_for_multi_routing(&MultipleNodeRoutingInfo::AllMasters) // TODO - this involves allocating a hash set and a vec. can this be avoided?
+            .into_iter()
+            .flat_map(|addr| self.connection_for_address(addr))
+    }
+
+    pub(crate) fn connection_for_identifier(&self, identifier: &Identifier) -> Option<Connection> {
+        let conn = self.connection_map.get(identifier)?;
+        Some(conn.clone())
+    }
+
+    pub(crate) fn connection_for_address(
+        &self,
+        address: &str,
+    ) -> Option<ConnectionAndIdentifier<Connection>> {
+        let identifier = Identifier(address.into());
+        let connection = self.connection_for_identifier(&identifier)?;
+        Some((identifier, connection))
+    }
+
+    pub(crate) fn address_for_identifier<'a, 'b: 'a>(
+        &'a self,
+        identifier: &'a Identifier,
+    ) -> Option<String> {
+        if self.connection_map.contains_key(identifier) {
+            Some(identifier.0.clone())
+        } else {
+            None
+        }
+    }
+
+    pub(crate) fn random_connections(
+        &self,
+        amount: usize,
+    ) -> impl Iterator<Item = ConnectionAndIdentifier<Connection>> + '_ {
+        self.connection_map
+            .iter()
+            .choose_multiple(&mut rand::thread_rng(), amount)
+            .into_iter()
+            .map(|(identifier, connection)| (identifier.clone(), connection.clone()))
+    }
+
+    pub(crate) fn replace_or_add_connection_for_address(
+        &mut self,
+        address: impl Into<String>,
+        connection: Connection,
+    ) -> Identifier {
+        let identifier = Identifier(address.into());
+        self.connection_map.insert(identifier.clone(), connection);
+        identifier
+    }
+
+    pub(crate) fn remove_connection(&mut self, identifier: &Identifier) -> Option<Connection> {
```

Comment:
so, if we remove the connection only from the connection map, but the node's address is still present in the slot map, and we have a command that routes to a slot that points to this node address.
So we'll try to get the connection from the address and find that we don't have it in. So we'll try to create a new connection to this node, and if it fails we will send the command to a random node, even though we already decided to remove this connection in the refresh_connections function. not for this PR but we need to understand how to handle changes only in the connection map and when and how it should trigger a slots refreshment

Reply:
once you have an identifier, a missing connection always means that a disconnect happened. That's one of the benefits of using strong typing - identifiers are only creatable by the connection_manager.

Reply:
currently it isn't true - we're searching for a connection based on its address, so a missing connection might say a disconnect happened or it's a new node. Am I wrong?

Reply:
no, the API says that you search for a connection based on an identifier. The user doesn't know that an identifier contains the address, and in future refactors, identifiers might not contain these addresses.
Of course, there's another problematic scenario - what happens if you have an existing identifier from an old container, but the connections container was replaced? The new container might not contain the address.